libs/vectorise/include/vectorise/uint/uint.hpp

```diff
@@ -39,338 +40,1012 @@ namespace vectorise {
+  : data_{other.data_}
+{}
+
+template <uint16_t S>
+constexpr UInt<S>::UInt(UInt &&other) noexcept
+  : data_{std::move(other.data_)}
+{}
+
+template <uint16_t S>
+constexpr UInt<S>::UInt(ContainerType const &&other)
+  : data_(std::move(other))
+{}
+
+template <uint16_t S>
+constexpr UInt<S>::UInt(WideContainerType const &&other)
+  : data_(std::move(other))
+{}
+
+template <uint16_t S>
+template <typename T>
+constexpr UInt<S>::UInt(T const &other, meta::IfIsAByteArray<T> *)
+{
+  std::fill(data_.wide.begin(), data_.wide.end(), 0);
+  std::copy(other.pointer(), other.pointer() + other.size(), data_.base.begin());
+}
+
+template <uint16_t S>
+template <typename T>
+constexpr UInt<S>::UInt(T const &other, meta::IfIsStdString<T> *)
+{
+  std::fill(data_.wide.begin(), data_.wide.end(), 0);
+  std::copy(other.begin(), other.end(), data_.base.begin());
+}
 
-    // TODO: Copy instead
-    *this = std::forward<T>(other);
+template <uint16_t S>
+template <typename T>
+constexpr UInt<S>::UInt(T const &number, meta::IfIsInteger<T> *)
+{
+  std::fill(data_.wide.begin(), data_.wide.end(), 0);
+  data_.wide[0] = static_cast<uint64_t>(number);
+}
+
+////////////////////////////
+/// assignment operators ///
+////////////////////////////
+
+template <uint16_t S>
+constexpr UInt<S> &UInt<S>::operator=(UInt const &v)
+{
+  data_.wide = v.data_.wide;
+  return *this;
+}
+
+template <uint16_t S>
+template <typename ArrayType>
+constexpr meta::HasIndex<ArrayType, UInt<S>> &UInt<S>::operator=(ArrayType const &v)
+{
+  std::fill(data_.wide.begin(), data_.wide.end(), 0);
+  std::copy(v.pointer(), v.pointer() + v.capacity(), data_.base.begin());
+
+  return *this;
+}
+
+template <uint16_t S>
+template <typename T>
+constexpr meta::HasNoIndex<T, UInt<S>> &UInt<S>::operator=(T const &v)
+{
+  std::fill(data_.wide.begin(), data_.wide.end(), 0);
+  data_.wide[0] = static_cast<WideType>(v);
+
+  return *this;
+}
+
+/////////////////////////////////////////////
+/// comparison operators for UInt objects ///
+/////////////////////////////////////////////
+
+template <uint16_t S>
+constexpr bool UInt<S>::operator==(UInt const &other) const
+{
+  bool ret = true;
+  for (std::size_t i = 0; i < WIDE_ELEMENTS; ++i)
+  {
+    ret &= (data_.wide[i] == other.elementAt(i));
   }
 
-  explicit UInt(uint64_t const &number)
+  return ret;
+}
+
+template <uint16_t S>
+constexpr bool UInt<S>::operator!=(UInt const &other) const
+{
+  return !(*this == other);
+}
+
+template <uint16_t S>
+constexpr bool UInt<S>::operator<(UInt const &other) const
+{
+  // Simplified version, as we're dealing with wider elements
+  for (std::size_t i = WIDE_ELEMENTS - 1; i >= 0; --i)
   {
-    union
-    {
-      uint64_t value;
-      uint8_t  bytes[sizeof(uint64_t) / sizeof(BaseType)];
-    } data;
-    data.value = number;
-
-    // FIXME: Assumes little endian
-    std::size_t i = 0;
-    for (; i < sizeof(uint64_t); ++i)
+    if (data_.wide[i] == other.elementAt(i))
```

Comment:
What happens if you compare two `UInt` which are equal?  It seems to me that the loop will run forever, but maybe I am wrong?

Reply:
hm, equality is at line 324 and it doesn't seem like an infinite loop.